prof: plumbing for continuous memory profiling

teskje opened the pull request:
This PR adds the necessary plumbing in Mz to support continuous memory profiling with Polar Signals.

* Extending the existing jemalloc profile parsing with parsing of memory mappings.
* Converting parsed profiles into the pprof Profile format.
* Exposing an HTTP endpoint to dump the pprof heap profile.

This PR also adds a feature flag that controls whether jemalloc profiling is activated or not. The default is still disabled. The tentative plan is to enable jemalloc profiling in staging for a while, then enable it in production, then remove the feature flag. This way we can move slowly and make sure there are no performance regressions.

TPC-H benchmark results are [here](https://github.com/MaterializeInc/cloud/issues/4590#issuecomment-1741004316), and they look reassuring.

### Motivation

  * This PR adds a known-desirable feature.

Part of https://github.com/MaterializeInc/cloud/issues/4590.

### Tips for reviewer

The commits can be reviewed separately.

Most of this PR is taken from @umanwizard's prior work in #16000. I rewrote it for my own understanding, adjusted details based on testing with the polarsignals-scraper, and added the feature flag (last commit).

A follow-up commit to the cloud repo will be necessary to set up the polarsignals-scraper.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/A

teskje commented on src/prof/src/pprof_types.proto:
```diff
@@ -0,0 +1,225 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+// This file was originally copied from
+//
+//    https://github.com/google/pprof/blob/main/proto/profile.proto
```
@benesch Can you officially confirm or deny that this is an appropriate way to include third-party code?

benesch replied:
Almost! I left a suggestion with the recommended format. The key points are:

  * The original copyright header ("Copyright 2016 Google Inc. All Rights Reserved.") must be retained.
  * Use a GitHub permalink (commit SHA rather than `main`) to reference the exact version of the file you downloaded.

umanwizard commented:
Thanks for reviving this!

It looks fine to me, but, I think we should do the following before landing it:

1. Ask QA to run the full test suite with the flag enabled.
2. Have some more eyes look at `all_build_ids` and carefully review it against the man pages for `man dl_iterate_phdr` and `man elf`. I wrote this function and IIRC Julian reviewed it, so we at least think/hope there are no bugs. But it is high-risk; as the doc comment suggests, it "is written in a hilariously unsafe way". It's already in the codebase, but it's dead code now. After this PR lands, it will become live unconditionally; that is, _not_ controlled by the feature flag.

teskje commented:
> Ask QA to run the full test suite with the flag enabled.

Good call! I think it makes sense to make the tests run with this flag enabled always. @MaterializeInc/testing What is the current way to enable an off-by-default feature flag always in CI?

> After this PR lands, it will become live unconditionally; that is, not controlled by the feature flag.

I'll give it a review.

I think it should be possible to have `all_build_ids` protected by the feature flag too. The `Lazy` value will be initialized on first access, so as long as we don't access `BUILD_IDS`, the unsafe code will not be executed. `BUILD_IDS` is only accessed by `parse_jeheap`, which in turn is invoked on results of `JemallocProfCtl::dump`. So if we make `dump` fail if profiling is deactivated, that will prevent `all_build_ids` from being executed if the flag is not set. Or is there a reason why we'd want `JemallocProfCtl::dump` to work even with profiling deactivated?

def- commented:
> https://github.com/orgs/MaterializeInc/teams/testing What is the current way to enable an off-by-default feature flag always in CI?

I'll fork this and do a run for that.

umanwizard commented:
> Or is there a reason why we'd want JemallocProfCtl::dump to work even with profiling deactivated?

No good reason that I can think of. We clear the profile on deactivation anyway, so it would just return nothing.

def- commented:
Yes, that's exactly what I did in    https://github.com/MaterializeInc/materialize/pull/22070/commits/8df2e400dafb151c39c53fa2c0ca991730b8cae8

teskje commented:
Cool! I added the `misc/python/materialize/mzcompose/__init__.py` change to this PR as well now.

def- reviewed (changes requested):
Weird, I thought with `enable_jemalloc_profiling` this should be enabled in CI, but coverage claims that much of the code is untested: https://buildkite.com/materialize/coverage/builds/241#_
I guess it might make sense because never actually save the pprof files, could we test that explicitly? It would not be great if saving a profile would cause a panic in production.

benesch commented on src/prof/src/pprof_types.proto:
```diff
@@ -0,0 +1,225 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+// This file was originally copied from
+//
+//    https://github.com/google/pprof/blob/main/proto/profile.proto
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
```suggestion
// Copyright 2016 Google Inc. All Rights Reserved.
// Copyright Materialize, Inc. and contributors. All rights reserved.
//
// This file is derived from the pprof project. The original source code was
// retrieved on September 29, 2023 from:
//
//    https://github.com/google/pprof/blob/f7f687d19a9829062f02a2dd6e9a6fb527dbf143/proto/profile.proto
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

umanwizard replied:
Isn’t line 2 of your suggestion factually wrong? We didn’t change anything in that file, so what would we own the copyright on?

benesch replied:
The lines that say "This file is derived from ..." and any future modifications to the file.

teskje commented:
All review comments should be addressed now:

* I have changed the HTTP endpoints that dump heap profiles to return an error if profiling is not activated. That means the `mz-build-id` code won't be executed if the `enable_jemalloc_profiling` feature flag is not set.
* I have reviewed the code of `mz-build-id`. I haven't found anything suspicious but opened a PR to add more asserts to be save (https://github.com/MaterializeInc/materialize/pull/22085).
* I have added a cluster test that queries the profiling endpoints and started a new coverage run.

ParkMyCar reviewed (approved):
The adapter bits LGTM, I did not really review any of the other parts

def- commented on test/cluster/mzcompose.py:
```diff
@@ -2546,3 +2548,83 @@ def worker(c: Composition, i: int) -> None:
+    c.up("materialized")
+    c.up("clusterd1")
+
+    def curl(container: str, *args: str) -> int:
```
I would use Python's requests library instead, will probably make errors easier to understand.

teskje replied:
Had to figure out how to get the right ports, but this is indeed nicer!

def- reviewed (approved):
Thanks for adding an explicit test and even checking coverage.

umanwizard commented:
@teskje Please let's not delete the annotation handling. It would be great to someday have per-thread allocation profiles.

teskje commented on test/cluster/mzcompose.py:
```diff
@@ -2546,3 +2548,83 @@ def worker(c: Composition, i: int) -> None:
+        user="mz_system",
+    )
+    code = curl("materialized", f"localhost:6878/prof/heap")
+    assert code == 200, f"expected 200, got {code}"
```
I added checking of distinctive substrings to the test.

teskje commented:
TFTRs!
